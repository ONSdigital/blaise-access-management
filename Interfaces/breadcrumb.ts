export interface BreadcrumbItem {
    link: string
    title: string
}

export interface BreadcrumbProps {
    BreadcrumbList: BreadcrumbItem[]
}