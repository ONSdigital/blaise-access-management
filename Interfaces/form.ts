import { FormEvent } from "react";
import { UserForm } from ".";

export interface CustomFormProps {
    initialValues?: { [key: string]: string }
    onSubmit?: (data: { [key: string]: string }) => void
    onReset?: () => FormEvent<HTMLFormElement>
    children: React.ReactNode
}

export interface CustomFormState {
    data: { [key: string]: string }
    validators: { [key: string]: ((val: string, name: string, formData: UserForm) => string[])[] | unknown }
    errors: { [key: string]: string[] }
}

export interface TextInputProps {
    placeholder?: string
    name: string
    value?: string
    label: string
    type?: string
    errors?: string[]
    onChange?: (val: string) => void
    validators: ((val: string, name: string, formData: UserForm) => string[])[]
    password?: boolean
}