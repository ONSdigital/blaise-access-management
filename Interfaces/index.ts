interface Instrument {
    installDate: string
    name: string
    expired: boolean
    serverParkName: string
    activeToday: boolean
    surveyDays: string[]
    link: string
    fieldPeriod: string
    surveyTLA: string
}

interface Survey {
    instruments: Instrument[]
    survey: string
}

export type { Instrument, Survey };

export interface ImportUser {
    name: string
    password: string
    role: string
    valid: boolean
    warnings: string[]
}

export interface UploadedUser {
    name: string
    created: boolean
}

export interface Role {
    name: string
    permissions: string[]
    description: string
}
