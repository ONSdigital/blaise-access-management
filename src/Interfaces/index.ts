export interface UploadedUser {
    name: string
    created: boolean
}

export interface ImportUser {
    name: string
    password: string
    role: string
    valid: boolean
    warnings: string[]
}

type JSONValue =
    | string
    | { [x: string]: JSONValue };

export interface JSONObject {
    [x: string]: JSONValue;
}

export interface UserForm {
    username?: string
    password?: string
    confirm_password?: string
}

export interface ValidatorType { name: string, validators: ((val: string, name: string, formData: { [key: string]: string }) => string[])[] }

export interface ContextProviderType {
    errors: {
        [key: string]: string[];
    };
    data: {
        [key: string]: string;
    };
    setFieldValue: (name: string, value: string) => void;
    registerInput: ({ name, validators }: ValidatorType) => () => void;
}

export interface BreadcrumbItem {
    link: string
    title: string
}

export interface BreadcrumbProps {
    BreadcrumbList: BreadcrumbItem[]
}