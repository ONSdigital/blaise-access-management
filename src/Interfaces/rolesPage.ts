import { UserRole } from "blaise-api-node-client";

export interface RolesTableProps {
    roles: UserRole[];
    listError: string;
}