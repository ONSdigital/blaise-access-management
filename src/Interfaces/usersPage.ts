import { User } from "blaise-api-node-client";
import { ImportUser, UploadedUser } from ".";

export interface UserRouteParams {
    user: string;
}

export interface UsersProps {
    currentUser: User | undefined;
}

export interface UsersTableProps {
    users: User[];
    currentUser: User | undefined;
    listError: string;
}

export interface SelectFileProps {
    setUsersToUpload: (users: ImportUser[]) => void;
    movePageForward: () => void;
}

export interface UsersToUploadSummaryProps {
    usersToImport: ImportUser[]
    uploadUsers: () => void
}

export interface UsersUploadedSummaryProps {
    usersUploaded: UploadedUser[]
    numberOfValidUsers: number
}

export interface ReturnPanel {
    visible: boolean
    message: string
    status: string
}