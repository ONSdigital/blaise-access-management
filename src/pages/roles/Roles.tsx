import React, {ReactElement, useEffect, useState} from "react";
import {getAllRoles} from "../../utilities/http";
import {ONSErrorPanel, ONSLoadingPanel, ONSPanel} from "blaise-design-system-react-components";
import Breadcrumbs from "../../Components/Breadcrumbs";
import RolesTable from "./RolesTable";
import {UserRole} from "blaise-api-node-client";

function Roles(): ReactElement {
    const [roles, setRoles] = useState<UserRole[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");
    const [listLoading, setListLoading] = useState<boolean>(true);

    useEffect(() => {
        getRolesList().then(() => {return;});
    }, []);

    async function getRolesList() {
        setRoles([]);
        setListLoading(true);

        const [success, roleList] = await getAllRoles();
        setListLoading(false);

        if (!success) {
            setListError("Unable to load roles.");
            return;
        }

        if (roleList.length === 0) {
            setListError("No installed roles found.");
        }

        setRoles(roleList);
    }

    return (
        <>
            <Breadcrumbs BreadcrumbList={
                [
                    {link: "/", title: "Home"}
                ]
            } />

            <main id="main-content" className="ons-page__main ons-u-mt-no">
                <h1 className="ons-u-mb-l">Manage roles</h1>

                <ONSPanel>To ensure consistency across environments, DST are responsible for creating and managing roles. If you require a new role, please reach out to DST for assistance.</ONSPanel>

                {listError.includes("Unable") && <ONSErrorPanel />}

                {
                    listLoading ?
                        <ONSLoadingPanel />
                        :
                        <RolesTable roles={roles} listError={listError} />
                }
            </main>
        </>
    );
}

export default Roles;
