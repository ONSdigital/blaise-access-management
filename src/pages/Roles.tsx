import React, {ReactElement, useEffect, useState} from "react";
import {Role} from "../../Interfaces";
import {ErrorBoundary} from "../Components/ErrorHandling/ErrorBoundary";
import {getAllRoles} from "../utilities/http";
import {ONSPanel} from "blaise-design-system-react-components";
import RoleList from "../Components/RoleList";

interface Panel {
    visible: boolean
    message: string
    status: string
}

function Roles(): ReactElement {
    const [panel, setPanel] = useState<Panel>({visible: false, message: "", status: "info"});

    const updatePanel = (visible = false, message = "", status = "info") => {
        setPanel(
            {visible: visible, message: message, status: status}
        );
    };

    const [roles, setRoles] = useState<Role[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");


    useEffect(() => {
        getRolesList().then(() => console.log("Call getRolesList Complete"));
    }, []);

    async function getRolesList() {
        setRoles([]);

        const [success, roleList] = await getAllRoles();

        if (!success) {
            setListError("Unable to load roles.");
            return;
        }

        console.log(roleList);

        if (roleList.length === 0) {
            setListError("No roles found.");
        }

        setRoles(roleList);
    }


    return (
        <>

            <ONSPanel hidden={!panel.visible} status={panel.status}>
                <p>{panel.message}</p>
            </ONSPanel>
            <ErrorBoundary errorMessageText={"Unable to load role table correctly."}>
                <RoleList list={roles}
                          listError={listError}/>
            </ErrorBoundary>
        </>
    );
}

export default Roles;
