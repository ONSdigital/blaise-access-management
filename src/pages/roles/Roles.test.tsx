import React from "react";
import { render, waitFor, cleanup, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { mock_server_request_Return_JSON } from "../../tests/utils";
import { act } from "react-dom/test-utils";
import { createMemoryHistory } from "history";
import { Router } from "react-router";
import { Role } from "../../../Interfaces";
import Roles from "./Roles";

const roleList: Role[] = [
    { name: "DST", permissions: ["Admin", "Bacon.access"], description: "A role" },
    { name: "BDSS", permissions: ["Admin"], description: "Another role" }
];

describe("Manage Roles page", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, roleList);
    });

    it("view users page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <Roles />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });

    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <Roles />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });

        await waitFor(() => {
            expect(screen.getByText(/Roles are created and managed by DST/i)).toBeDefined();
            expect(screen.getByText(/Manage roles/i)).toBeDefined();
            expect(screen.getAllByText(/DST/i).length).toBeGreaterThan(1);
            expect(screen.getByText(/BDSS/i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});


describe("Given the API returns malformed json", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, { text: "Hello" });
    });

    it("it should render with the error message displayed", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <Roles />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });


        await waitFor(() => {
            expect(screen.getByText(/Sorry, there is a problem with this service. We are working to fix the problem. Please try again later./i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Given the API returns an empty list", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, []);
    });

    it("it should render with a message to inform the user in the list", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <Roles />
            </Router>
        );

        await act(async () => {
            await new Promise(process.nextTick);
        });


        await waitFor(() => {
            expect(screen.getByText(/No installed roles found./i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
