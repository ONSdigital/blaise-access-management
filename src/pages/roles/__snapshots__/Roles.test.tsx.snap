// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Roles page view users page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="u-mt-m"
      >
        <a
          href="/"
        >
          Previous
        </a>
      </p>
      <h1
        class="u-mt-s"
      >
        Manage roles
      </h1>
      <ul
        class="list list--bare list--inline "
      >
        <li
          class="list__item "
        >
          <a
            href="/roles/new"
          >
            Create new role
          </a>
        </li>
      </ul>
      <table
        class="table u-mt-m"
        id="roles-table"
      >
        <thead
          class="table__head"
        >
          <tr
            class="table__row"
          >
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Name
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Description
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Number of permissions
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body"
        >
          <tr
            class="table__row"
            data-testid="user-table-row"
          >
            <td
              class="table__cell "
            >
              DST
            </td>
            <td
              class="table__cell "
            >
              A role
            </td>
            <td
              class="table__cell "
            >
              2
            </td>
          </tr>
          <tr
            class="table__row"
            data-testid="user-table-row"
          >
            <td
              class="table__cell "
            >
              BDSS
            </td>
            <td
              class="table__cell "
            >
              Another role
            </td>
            <td
              class="table__cell "
            >
              1
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <p
      class="u-mt-m"
    >
      <a
        href="/"
      >
        Previous
      </a>
    </p>
    <h1
      class="u-mt-s"
    >
      Manage roles
    </h1>
    <ul
      class="list list--bare list--inline "
    >
      <li
        class="list__item "
      >
        <a
          href="/roles/new"
        >
          Create new role
        </a>
      </li>
    </ul>
    <table
      class="table u-mt-m"
      id="roles-table"
    >
      <thead
        class="table__head"
      >
        <tr
          class="table__row"
        >
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Name
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Description
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Number of permissions
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        class="table__body"
      >
        <tr
          class="table__row"
          data-testid="user-table-row"
        >
          <td
            class="table__cell "
          >
            DST
          </td>
          <td
            class="table__cell "
          >
            A role
          </td>
          <td
            class="table__cell "
          >
            2
          </td>
        </tr>
        <tr
          class="table__row"
          data-testid="user-table-row"
        >
          <td
            class="table__cell "
          >
            BDSS
          </td>
          <td
            class="table__cell "
          >
            Another role
          </td>
          <td
            class="table__cell "
          >
            1
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
